local_pathfinding/obstacles.py

```diff
@@ -0,0 +1,340 @@
+                speed_over_ground_kmph * np.cos(math.radians(course_over_ground)),
+            ]
+        )
+
+        time_to_intersection = self.calculate_time_to_intersection(
+            position, boat_sog_vector, sailbot_position, sailbot_speed
+        )
+
+        # Sailbot and this Boat will never collide
+        if time_to_intersection < 0:
+            return PROJ_TIME_NO_COLLISION * speed_over_ground_kmph
+
+        # Added 1 km to the projected distance for extra safety margin
+        return time_to_intersection * speed_over_ground_kmph + 1
+
+    def calculate_time_to_intersection(
+        self,
+        position: XY,
+        boat_sog_vector: NDArray,
+        sailbot_position: XY,
+        sailbot_speed: float,
+    ) -> float:
+        """
+        Calculates the time until the boat and Sailbot collide, if the Sailbot moves
+        directly towards the soonest possible collision point at its current speed.
+        The system is modeled by two parametric lines extending from the positions of the boat
+        obstacle and sailbot respectively in 2D space. These lines may intersect at some specific
+        point and time.
+
+        This linear system, in which the vector that represents the Sailbot's velocity is free
+        to point at the soonest possible collision point (but whose magnitude is known),
+        is solved using linear algebra and  the quadratic formula.
+
+        A more in-depth explanation for this function can be found here:
+        https://ubcsailbot.atlassian.net/wiki/spaces/prjt22/pages/1881145358/Obstacle+Class+Planning
+
+
+        Args:
+            position (XY): x,y coordinates of the boat in km
+            boat_sog_vector (np.array): x,y components of the boat's speed over ground
+            sailbot_position (XY): x,y coordinates of the Sailbot in km
+        Returns:
+            time_to_intersection (float): time in hours until the boat and Sailbot collide
+            -1 if the boats will never collide
+        """
+        v1 = boat_sog_vector[0]
+        v2 = boat_sog_vector[1]
+        a = position[0]
+        b = position[1]
+        c = sailbot_position[0]
+        d = sailbot_position[1]
+
+        quadratic_coefficients = np.array(
+            [
+                v1**2 + v2**2 - (sailbot_speed**2),
+                2 * (v1 * (a - c) + v2 * (b - d)),
+                (a - c) ** 2 + (b - d) ** 2,
+            ]
+        )
+
+        # If the radicand  of the quadratic formula is negative, the boats will never collide
+        if (2 * (v1 * (a - c) - v2 * (b - d))) ** 2 - 4 * (
+            v1**2 + v2**2 - (sailbot_speed**2)
+        ) * (a - c) ** 2 + (b - d) ** 2 <= 0:
+            return -1
+
+        # The solution to the quadratic formula is the time until the boats collide
+        t = np.roots(quadratic_coefficients)
+
+        # Return the smaller positive time
+        return min([i for i in t if i > 0])
```

Comment:
Is it possible that there could be no element > 0? If so, `min` would raise a `ValueError`

Reply:
Yes, that is possible. I will add a try except block to check for that and return -1 if there are no positive elements. 

Generally, its ok if there doesn't end up being any positive times, just means they cant collide with the current parameters. So projecting the Boat's position up to the maximum distance would be the appropriate action in that case.